### PR TITLE
logind: Add new flag for kexec reboot

### PR DESCRIPTION
**Pull request by deepak-rawat**
Add new flag to allow kexec reboot if kernel is already loaded.

### CODE REVIEW

**Comment by deepak-rawat** [comment]
I think better option will be to use a flag like SD_LOGIND_KEXEC_REBOOT and only allow kexec reboot with RebootWithFlags. Would that be acceptable?

**Review by bluca — APPROVED** [review]
This matches the behaviour of 'systemctl reboot', so LGTM.

I think we can include it in v248, since it's good to have it together with the new RebootWithFlags logind interface

**Comment by bluca** [comment]
> I think better option will be to use a flag like SD_LOGIND_KEXEC_REBOOT and only allow kexec reboot with RebootWithFlags. Would that be acceptable?

systemctl does this implicitly, so that doesn't seem necessary to me - but no strong objections

**Comment by poettering** [comment]
what's the rationale for this? this is quite some behaviour change...

so far any kexec_loaded() check that changed reboot to kexec were done client side? What is the background for moving this server side? And if so, why keep the client side in place? this shouldn't be duplicated on both sides

**Comment by poettering** [comment]
> I think we can include it in v248, since it's good to have it together with the new RebootWithFlags logind interface

Let's not hurry this in. The situation is far from clear to me here...

**Comment by bluca** [comment]
> what's the rationale for this? this is quite some behaviour change...
> 
> so far any kexec_loaded() check that changed reboot to kexec were done client side? What is the background for moving this server side? And if so, why keep the client side in place? this shouldn't be duplicated on both sides

There's no way to do a kexec reboot via logind (needed because of the inhibitors feature) through DBUS, currently. This attempts to fill that gap.

If we want to keep implicit behaviour changes only client-side, then a new flag as Deepak suggested would be better?

> Let's not hurry this in. The situation is far from clear to me here...

Ack

**Comment by bluca** [comment]
By new flag here we mean for the RebootWithFlags() DBUS method

**Comment by poettering** [comment]
I mean, other option would be to add a high-level KExec method call to logind. Not saying this would be better, but it's an option to think about.

Maybe the right approach is to drop all redirecting-to-kexec from the client-side, and move it into logind inside, but make it controllable by a new flag. All our clients would set the flag, so that redirect-to-kexec would be the default. And then we'd add an env var to turn this flag the clients set off again. dunno?

**Comment by bluca** [comment]
> I mean, other option would be to add a high-level KExec method call to logind. Not saying this would be better, but it's an option to think about.

org.freedesktop.systemd1 has separate Reboot()/KEexec() methods, so if we want consistency across interfaces this could be the way to go.

> Maybe the right approach is to drop all redirecting-to-kexec from the client-side, and move it into logind inside, but make it controllable by a new flag. All our clients would set the flag, so that redirect-to-kexec would be the default. And then we'd add an env var to turn this flag the clients set off again. dunno?

Thinking about this a bit more, this would factually be a change in behaviour of this stable API interface, so I'm leaning toward adding a new method to logind's interface.

**Comment by poettering** [comment]
> org.freedesktop.systemd1 has separate Reboot()/KEexec() methods, so if we want consistency across interfaces this could be the way to go.

PID 1 is a lot more low-level than logind. So we definitely should expose separate calls on the low-level PID1. But we can do something different in logind, with more automatic magic

**Comment by bluca** [comment]
> > org.freedesktop.systemd1 has separate Reboot()/KEexec() methods, so if we want consistency across interfaces this could be the way to go.
> 
> PID 1 is a lot more low-level than logind. So we definitely should expose separate calls on the low-level PID1. But we can do something different in logind, with more automatic magic

Ok, sounds good to me.

@deepak-rawat could you please push a new revision that adds a flag to disable the redirect, and drop the redirect from systemctl?

**Comment by deepak-rawat** [comment]
Looked into client side and only "/usr/bin/reboot" and /usr/bin/shutdown -r" support kexec reboot if kernel is loaded (/sys/kernel/kexec_loaded is 1). Executing "systemctl reboot" won't result in kexec reboot, even if kernel is loaded. So I think best is to introduce a new flag "SD_LOGIND_KEXEC_REBOOT" which is applicable to RebootWithFlags method only. If this is set then perform a kexec reboot. On client side add support for ACTION_KEXEC with logind_reboot.

**Review by bluca — APPROVED** [review]
lgtm

**Comment by bluca** [comment]
centos failed on git clone (ping @mrc0mmand for a restart if possible - thanks) and bionic is the known issue

File: man/org.freedesktop.login1.xml [hunk 1]
```diff
@@ -555,10 +555,13 @@ node /org/freedesktop/login1 {
       extendability, defined as follows:</para>
       <programlisting>
 #define SD_LOGIND_ROOT_CHECK_INHIBITORS  (UINT64_C(1) &lt;&lt; 0)
+#define SD_LOGIND_KEXEC_REBOOT           (UINT64_C(1) &lt;&lt; 1)
```
Comment (poettering) [inline man/org.freedesktop.login1.xml:558]: I think SD_LOGIND_REBOOT_BY_KEXEC or SD_LOGIND_REBOOT_VIA_KEXEC or so better, to make clear this is about redirecting operations

File: src/login/logind-dbus.c [hunk 2]
```diff
@@ -1880,6 +1880,8 @@ static int method_do_shutdown_or_sleep(
                         return r;
                 if ((flags & ~SD_LOGIND_SHUTDOWN_AND_SLEEP_FLAGS_PUBLIC) != 0)
                         return sd_bus_error_setf(error, SD_BUS_ERROR_INVALID_ARGS, "Invalid flags parameter");
+                if (!streq(unit_name, SPECIAL_REBOOT_TARGET) && (flags & SD_LOGIND_KEXEC_REBOOT))
+                        return sd_bus_error_setf(error, SD_BUS_ERROR_INVALID_ARGS, "Invalid flags parameter");
```
Comment (poettering) [inline src/login/logind-dbus.c:1884]: this should return a proper error message, i.e. "reboot-via-kexec functionality is only available for reboot operations", or so

**Pull request merged**